### ISSUE TITLE
Push down ScopeMappable marker interface from AbstractField to AbstractWindowFunction and TableFieldImpl

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14730]
`DefaultRenderContext::scopeMapping` has been used so far mainly for 2 purposes:

- To register implicit joins *per scope* (see https://github.com/jOOQ/jOOQ/issues/1502)
- To register `QUALIFY` emulated window function expressions *per scope* (see https://github.com/jOOQ/jOOQ/issues/5810)

Both features are used relatively rarely (especially the latter), meaning that the overhead of looking up things in the `ScopeStack` should be avoided whenever possible. The lookup is usually not *that* expensive, but it can happen that `JSONB` content needs to be parsed because of https://github.com/jOOQ/jOOQ/issues/12023:

```
java.lang.Throwable
	at org.jooq.JSONB.parsed(JSONB.java:102)
	at org.jooq.JSONB.hashCode(JSONB.java:116)
	at org.jooq.impl.AbstractParam.hashCode(AbstractParam.java:278)
	at java.base/java.util.HashMap.hash(HashMap.java:338)
	at java.base/java.util.HashMap.getNode(HashMap.java:568)
	at java.base/java.util.LinkedHashMap.get(LinkedHashMap.java:441)
	at org.jooq.impl.ScopeStack.listOrNull(ScopeStack.java:233)
	at org.jooq.impl.ScopeStack.get(ScopeStack.java:198)
	at org.jooq.impl.DefaultRenderContext.scopeMapping(DefaultRenderContext.java:204)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:292)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:260)
	at org.jooq.impl.Eq.lambda$2(Eq.java:104)
	at org.jooq.impl.Eq.acceptCompareCondition(Eq.java:155)
	at org.jooq.impl.Eq.accept(Eq.java:104)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:726)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:350)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:255)
	at org.jooq.impl.ConditionAsField.accept(ConditionAsField.java:129)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:726)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:301)
	at org.jooq.impl.QueryPartCollectionView.acceptElement(QueryPartCollectionView.java:251)
	at org.jooq.impl.SelectFieldList.lambda$1(SelectFieldList.java:93)
	at org.jooq.impl.Tools.visitAutoAliased(Tools.java:2811)
	at org.jooq.impl.SelectFieldList.acceptElement0(SelectFieldList.java:93)
	at org.jooq.impl.SelectFieldList.acceptElement(SelectFieldList.java:89)
	at org.jooq.impl.SelectFieldList.acceptElement(SelectFieldList.java:1)
	at org.jooq.impl.QueryPartCollectionView.accept(QueryPartCollectionView.java:223)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:726)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:350)
	at org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:2327)
	at org.jooq.impl.SelectQueryImpl.lambda$36(SelectQueryImpl.java:1981)
	at org.jooq.impl.AbstractContext.toggle(AbstractContext.java:393)
	at org.jooq.impl.AbstractContext.data(AbstractContext.java:404)
	at org.jooq.impl.SelectQueryImpl.toSQLReferenceLimitDefault(SelectQueryImpl.java:1981)
	at org.jooq.impl.SelectQueryImpl.accept0(SelectQueryImpl.java:1876)
	at org.jooq.impl.SelectQueryImpl.accept(SelectQueryImpl.java:1504)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:726)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:350)
	at org.jooq.impl.DefaultDSLContext.render(DefaultDSLContext.java:734)
	at org.jooq.impl.AbstractAttachableQueryPart.getSQL(AbstractAttachableQueryPart.java:121)
	at org.jooq.impl.AbstractAttachableQueryPart.getSQL(AbstractAttachableQueryPart.java:114)
	at org.jooq.impl.AbstractDelegatingQuery.getSQL(AbstractDelegatingQuery.java:95)
	at org.jooq.test.benchmarks.ci.BenchmarkRenderingJSONBBinds.testRender(BenchmarkRenderingJSONBBinds.java:52)
	at org.jooq.test.benchmarks.ci.BenchmarkRenderingJSONBBinds.main(BenchmarkRenderingJSONBBinds.java:64)
```

At the time of https://github.com/jOOQ/jOOQ/issues/5810, `AbstractField` was made `ScopeMappable`, which may have been a bit blunt, given that this emulation only ever puts `AbstractWindowFunction` types (and at that, those that also return `true` for `isWindow()`) in the `ScopeStack`.

**Comment by lukaseder** [comment]
It seems there's only a marginal improvement:

```
Benchmark                                   Mode  Cnt       Score       Error  Units
BenchmarkRenderingJSONBBinds.testRender    thrpt    7  365519.152 ± 18392.821  ops/s
BenchmarkRenderingJSONBBinds.testToString  thrpt    7  390925.374 ± 17150.308  ops/s


Benchmark                                   Mode  Cnt       Score       Error  Units
BenchmarkRenderingJSONBBinds.testRender    thrpt    7  383613.905 ± 26020.586  ops/s
BenchmarkRenderingJSONBBinds.testToString  thrpt    7  401023.618 ± 19004.640  ops/s
```

It's still worth it, but given the feedback from here https://github.com/jOOQ/jOOQ/issues/14721, I would have expected more.

**Comment by lukaseder** [comment]
Ah, #14721 generates this overhead when *constructing* the SQL, not when rendering it.

**Comment by lukaseder** [comment]
Unfortunately, this also produces a regression in one of the emulation tests.

**Comment by lukaseder** [comment]
Although, for an `UPDATE` query, the improvement is quite significant (if paired with lazy evaluation of `Param` name):

**Before**

```
Benchmark               Mode  Cnt       Score       Error  Units
JSONBBinds.testRender  thrpt    7  181476.742 ± 12590.682  ops/s
```

**After**

```
Benchmark               Mode  Cnt       Score      Error  Units
JSONBBinds.testRender  thrpt    7  386594.927 ± 4703.305  ops/s
```

**Comment by lukaseder** [comment]
I misunderstood how the `ScopeMappable` feature was used to emulate `QUALIFY`. It isn't required on the `AbstractWindowFunction`, but on the `TableFieldImpl` for disambiguation of columns in the derived table.

This disambiguation is needed a few times in jOOQ, so perhaps it can be re-implemented more generically. In any case, let's try this again with the push down. In addition to that, more regression tests need to be added for the `QUALIFY` emulation, see also: https://github.com/jOOQ/jOOQ/issues/12098

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.9 (https://github.com/jOOQ/jOOQ/issues/14734), and 3.16.15 (https://github.com/jOOQ/jOOQ/issues/14735)

**Comment by lukaseder** [comment]
This pushdown might have introduced a regression when `FieldProxy` is involved? https://github.com/jOOQ/jOOQ/issues/14890
